### PR TITLE
Add `load_dataset` implementation

### PR DESCRIPTION
**Pull request by goeffthomas**
This adds the `load_dataset` implementation without including it to the exported `kagglehub` module yet. Before adding to the `kagglehub` export (and making the feature more easily discoverable), I want to add more testing around all the different file types and settings that are possible to load. This will come on a follow up PR along with added documentation on how to use it.

Notes:
- This PR also includes adding the `hf-datasets`/`pandas-datasets` optional dependencies and hatch environments to install them
- The implementation for this feature has changed a few times. Those other iterations should be visible in the commit history, but as a recap:
  - `apache_beam` doesn't play nicely with `datasets` due to an underlying dependency issue
  - `mlcroissant` (and Croissant itself) has some major file type limitations
  - Curated/validated arguments for `datasets` and `pandas`
  - **Current:** Simple passthrough `kwargs` for the underlying dependencies
- As implemented, memory is the main constraint. [dask](https://docs.dask.org/en/stable/dataframe.html) looks like a great way to unlock more/larger datasets, but I'm leaning toward leaving that as an improvement if users run into the in-memory problems. Open to feedback on this decision (or any others big/small re: the implementation).

I went a little wide with reviews, but this is feeling more like a direction I could see us going and want to make sure I'm not missing any blind spots. This is my first major python contribution, so don't be shy about feedback 🙏 

http://b/379756505

### CODE REVIEW

File: src/kagglehub/clients.py [hunk 1]
```diff
@@ -198,6 +201,24 @@ def download_file(
                         _CHECKSUM_MISMATCH_MSG_TEMPLATE.format(expected_md5_hash, actual_md5_hash)
                     )
 
+            # For individual file downloads, the downloaded file may be a zip of the file rather
```
Comment (neshdev) [inline src/kagglehub/clients.py:204]: Can you split the bugfix to its own PR? Can you also include a unit test for the fix?

File: tools/scripts/download_model.py
Comment (neshdev) [inline tools/scripts/download_model.py:1]: Can we add a readme file for the scripts. Just something to ack that these are just examples and are for testing.

File: stubs/datasets/__init__.pyi [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import pandas as pd
+
+class Dataset:
+    @classmethod
+    def from_pandas(cls, df: pd.DataFrame) -> Dataset: ...
+
+class DatasetDict(dict): ...
```
Comment (neshdev) [inline stubs/datasets/__init__.pyi:7]: I have not used stubs before but seems we can add typing to other libs using stubs. What is this a stub for?

File: src/kagglehub/pandas_datasets.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+# list of `pandas-datasets` optional dependencies in pyproject.toml.
+import pandas as pd
+
+from kagglehub.datasets import dataset_download
+
+
+# This is a thin wrapper around pd.read_sql_query so we get the connection
+# closing for free after we've produced the DataFrame
+def wrapped_read_sql_query(sql_query: str, path: str) -> pd.DataFrame:
+    conn = sqlite3.connect(path)
```
Comment (neshdev) [inline src/kagglehub/pandas_datasets.py:18]: Is it possible to use a with statement here so the resource gets closed regardless of an error?

**Review by neshdev — APPROVED** [review]
LGTM

File: stubs/datasets/__init__.pyi (hunk 2 above)
Comment (goeffthomas) [reply stubs/datasets/__init__.pyi:7]: It's probably not clear because HF's python package for their datasets is actually just `datasets`. This are stubs for the classes we use. I had failing lint checks from mypy without this.

File: docker-hatch [hunk 4]
```diff
@@ -34,6 +34,14 @@ while :; do
         -f|--force-rebuild-docker-image)
             CACHE_FLAG='--no-cache'
             ;;
+        # These are flags for hatch commands: https://hatch.pypa.io/latest/cli/reference/#hatch
```
Comment (rosbo) [inline docker-hatch:37]: I don't think this is required, the way `docker-hatch` work is that it looks at its own flag until we reach the hatch command. It then break from the loop and pass the rest as-is.

If you run:

```
./docker-hatch run -h
```

Then, under the hood it will run: `hatch run -h`.

The only case you can't use hatch flag is if you try to run a hatch without a command.

Say you wanted to run:

```
hatch --help
```

If you run:

```
./docker-hatch --help
```

Then, the `--help` flag will be read by the script. I would simply remove line 21-24 in this script.

File: README.md [hunk 5]
```diff
@@ -205,11 +205,34 @@ hatch test integration_tests
...
 ### Run `kagglehub` from source
 
+#### Option 1: Execute a one-liner of code from the command line
+
 ```sh
 # Download a model & print the path
 hatch run python -c "import kagglehub; print('path: ', kagglehub.model_download('google/bert/tensorFlow2/answer-equivalence-bem'))"
 ```
 
+#### Option 2: Run a saved script from the /tools/scripts directory
```
Comment (rosbo) [inline README.md:215]: Can you open a separate PR to add this feature: [go/small-cls](https://goto.google.com/small-cls)

File: src/kagglehub/datasets.py [hunk 6]
```diff
@@ -60,3 +72,65 @@ def dataset_upload(
     )
 
     create_dataset_or_version(h, tokens, version_notes)
+
+
+def load_dataset(
+    adapter: KaggleDatasetAdapter,
```
Comment (rosbo) [inline src/kagglehub/datasets.py:78]: Do we want a default value or do we prefer always having the user explicitly set what they want?

Maybe explicit is better, the user will know what they will get.

File: src/kagglehub/datasets.py [hunk 7]
```diff
@@ -60,3 +72,65 @@ def dataset_upload(
...
+    # In the form of {owner_slug}/{dataset_slug} or {owner_slug}/{dataset_slug}/versions/{version_number}
+    handle: str,
+    path: str,
+    *,
+    # Kaggle doesn't currently support test/train splits natively like Hugging Face does, so users
+    # would specify here if they select the 'hugging_face' adapter.
```
Comment (rosbo) [inline src/kagglehub/datasets.py:84]: We should use **kwargs and just pass that to the underlying method. This way,  can support everything the underlying method supports (e.g. pd.read_csv).

File: docker-hatch (hunk 4 above)
Comment (goeffthomas) [reply docker-hatch:37]: Maybe I can give this another look, but I couldn't find a way to run with the `-v` flag to get verbose logging while creating environments which was crucial for finding where the pip installs were hanging. The -v was getting swallowed up by our `-v` to use python version. Maybe you're right and I just needed to rename ours to `-P` and the rest would work. I can try again to see if that alone does the job and then possibly remove this.

File: src/kagglehub/datasets.py (hunk 6 above)
Comment (goeffthomas) [reply src/kagglehub/datasets.py:78]: Yeah, because we now have 2 to start (instead of just the one HF one), I preferred explicit.

File: src/kagglehub/datasets.py (hunk 7 above)
Comment (goeffthomas) [reply src/kagglehub/datasets.py:84]: I was a bit worried about how this scales, so I'll give this a try. I think my thought was that it'd be nice to validate upfront before triggering a user download, but maybe that pushes too much maintenance cost onto us.

File: src/kagglehub/hf_datasets.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    # This shouldn't happen since we validated the sheet_name argument already, but we can check again to
+    # be robust and enable type safety in the else block.
+    if isinstance(result, dict):
+        raise ValueError(MULTIPLE_SHEETS_ERROR_MESSAGE)
+    else:
+        # Split the DataFrame into train/test splits based on the user-specified split
+        split_index = int(result.shape[0] * train_split_percent)
+        train_split = Dataset.from_pandas(result[:split_index])
+        test_split = Dataset.from_pandas(result[split_index:])
```
Comment (pcuenca) [inline src/kagglehub/hf_datasets.py:59]: Shouldn't this default to random? For example, in the dataset used as an example in `load_hf_dataset.py`, entries are sorted in reverse order. Or use [`train_test_split`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.train_test_split) and somehow expose similar options, as I assume many people would want control over the splits.

File: stubs/datasets/__init__.pyi (hunk 2 above)
Comment (neshdev) [reply stubs/datasets/__init__.pyi:7]: We haven't used stubs for all our other modules and seems non standard for this project. Can you just move the typing info into `datasets.py`

File: docker-hatch (hunk 4 above)
Comment (rosbo) [reply docker-hatch:37]: I don't think you need to rename our `-v` flag to `-P`. As long as you pass the hatch `-v` flag after whatever command you wanted to run, it should work.

```
./docker-hatch -v 3.9 COMMAND -v
```

The command above would run `hatch command -v` in a Python 3.9 docker image.

Comment (goeffthomas) [reply docker-hatch:37]: Ah, right, I remember now. The problem is our case statement below that flags unknown options and prints the `usage` and exits. So either we put the known hatch flags in here or we remove that block. Maybe the latter is better.

Comment (rosbo) [reply docker-hatch:37]: as long as you put the hatch flags after the hatch COMMAND, it should just work.

Can you share the command you are trying to run?

File: stubs/datasets/__init__.pyi (hunk 2 above)
Comment (neshdev) [reply stubs/datasets/__init__.pyi:7]: LGTM - we resolve this offline. @goeffthomas will add comments to the file explaning that this stub is for the HF package called `datasets`.

File: docker-hatch (hunk 4 above)
Comment (goeffthomas) [reply docker-hatch:37]: `./docker-hatch -v env create hf-datasets`

If you put the `-v` after `env`, you get `Error: No such option` because `-v` isn't a flag for the `env` command`, it's a flag for `hatch`.

Comment (rosbo) [reply docker-hatch]: Ah. I see, now I get it. The way `blaze` handle this is with the `--`:

```
blaze run cloud/kaggle/kernels/tools:vmtd -- --binary $ABUSE_BINARY_MD5HASH
```

We could do:

```
./docker-hatch -- -v env create hf-datasets.
```

And in our logic reading flag in a loop, break if we get `--`.

File: src/kagglehub/hf_datasets.py (hunk 8 above)
Comment (goeffthomas) [reply src/kagglehub/hf_datasets.py]: I've updated the implementation pretty significantly. **tl;dr** We're just passing `kwargs` through to the underlying dependencies. So now you just get a `Dataset` and can do anything you want with it (like `train_test_split`, etc.). So users get the easy one-liner to load a HF Dataset from Kaggle, but then all the full-fledged features they want/enjoy from HF.

File: src/kagglehub/hf_datasets.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    # We're only loading one Dataset at a time, so alert the user if their args produce multiple. This will
+    # probably only happen if the user is trying to load an Excel-based file with many sheets and none specified
+    if isinstance(result, dict) or isinstance(result, list):
+        raise ValueError(MULTIPLE_DATA_FRAMES_ERROR_MESSAGE)
+
+    # NOTE: Order matters here as we're letting users override our specified defaults, namely preserve_index=False.
+    # This may be valuable in the edge case that a user does actually want the index persisted as a column.
+    merged_kwargs = {**DEFAULT_PANDAS_KWARGS, **hf_kwargs}
+    return Dataset.from_pandas(result, **merged_kwargs)
```
Comment (rosbo) [inline src/kagglehub/hf_datasets.py:55]: Can you call `load_dataset` directly instead? https://screenshot.googleplex.com/552iQEq9vcFF2V9

And not split kwargs between pandas & hf?

Comment (goeffthomas) [reply src/kagglehub/hf_datasets.py:55]: If we do that, we'd lose support for sqlite, Excel-like files, and feather

**Review by rosbo — APPROVED** [review]

**Comment by goeffthomas** [comment]
@jplotts @neshdev I'm going to merge this now because I have a follow up PR ready to open. I can apply any feedback you have here to that follow up PR.
